### PR TITLE
Accumulate 16-bit float sums in 32-bit accumulators

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60389 WIP: Accumulate float sums in double accumulators
* **#60387 Accumulate 16-bit float sums in 32-bit accumulators**

Fixes gh-59489

Using 32-bit accumulators is a win-win: improved precision and improved
performance since the half precision types needed to be converted back and forth
to 32-bit float to do the arithmetic anyway.

Note that on multi-threaded or dis-contiguous sums, there can be partial sums
stored in the output so they are necessarily trucated to 16-bit. Fixing this
would require a rework of TensorIterator reductions.

Differential Revision: [D29447187](https://our.internmc.facebook.com/intern/diff/D29447187)

### CODE REVIEW

**Comment by peterbell10** [comment]
MacOS failure appears to be unrelated, probably due to a change in behaviour of the recently released SciPy 1.7 (scipy/scipy#13362).

**Comment by ngimel** [comment]
This looks good, can you post before and after benchmarks for fp32 (to make sure it doesn't regress) and bfloat16?

**Comment by peterbell10** [comment]
I've benchmarked with this simple `timeit` loop:
```python
import torch
torch.set_num_threads(1)
for dtype in [torch.float, torch.bfloat16]:
    print(dtype)
    for i in range(1, 6):
        a = torch.rand(10**i, 1000, dtype=dtype)
        print(a.shape)
        %timeit a.sum(0)
        %timeit a.sum(1)
        b = torch.rand(10**i, 2000, dtype=dtype)[:, ::2]
        %timeit b.sum(0)
        %timeit b.sum(1)
```


#### Float results
Float performance is the same, to within measurement error

|        Shape | Dim | Master Contiguous (us) | This PR Contiguous (us) | Master Discontiguous (us) | This PR Discontiguous (us) |
|-------------:|-----|:----------------------:|:-----------------------:|:-------------------------:|:--------------------------:|
|     10, 1000 | 0   |          1.92          |           1.92          |            4.33           |            4.34            |
|              | 1   |          1.71          |           1.72          |            2.86           |            2.93            |
|    100, 1000 | 0   |          6.42          |           6.84          |            27.7           |            27.3            |
|              | 1   |          5.65          |           6.01          |            17.2           |            17.1            |
|   1000, 1000 | 0   |           57           |           56.2          |            259            |             256            |
|              | 1   |          41.7          |           41.5          |            150            |             152            |
|  10000, 1000 | 0   |          1,360         |          1,340          |           8,020           |            7,890           |
|              | 1   |           838          |           823           |           3,030           |            2,970           |
| 100000, 1000 | 0   |         24,900         |          24,200         |           92,200          |           91,400           |
|              | 1   |         12,100         |          11,800         |           30,900          |           31,100           |

#### BFloat16 results
For BFloat16 this is consistently around 3x faster

|        Shape | Dim | Master Contiguous (us) | This PR Contiguous (us) | Master Discontiguous (us) | This PR Discontiguous (us) |
|-------------:|-----|:----------------------:|:-----------------------:|:-------------------------:|:--------------------------:|
|     10, 1000 | 0   |          5.19          |           3.51          |            17.6           |            7.29            |
|              | 1   |          4.15          |           2.00          |            13.4           |            4.39            |
|    100, 1000 | 0   |          22.7          |           7.02          |            124            |            40.9            |
|              | 1   |          29.1          |           8.02          |            122            |            31.6            |
|   1000, 1000 | 0   |           213          |           50.1          |           1,190           |             404            |
|              | 1   |           283          |           68.2          |           1,210           |             303            |
|  10000, 1000 | 0   |          3,760         |           921           |           14,500          |            9,230           |
|              | 1   |          2,860         |           679           |           12,200          |            3,110           |
| 100000, 1000 | 0   |         82,500         |          20,700         |          151,000          |           99,100           |
|              | 1   |         29,400         |          8,590          |          122,000          |           31,900           |

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29447187).

**Comment by ngimel** [comment]
Internal builds are failing with 
```
caffe2/aten/src/ATen/native/cpu/SumKernel.cpp:351:17: error: unused variable 'vec_size' [-Werror,-Wunused-variable]
  const int64_t vec_size = size0 / vacc_t::size();
```

File: aten/src/ATen/native/cpu/SumKernel.cpp [hunk 1]
```diff
@@ -220,39 +341,42 @@ void vectorized_outer_sum(
...
     int64_t size0, int64_t size1) {
   using vec_t = Vectorized<scalar_t>;
+  using acc_t = at::acc_type<scalar_t, true>;
+  using vacc_t = Vectorized<acc_t>;
+  using VecLoadPolicy = OuterSumCastLoadPolicy<vec_t, vacc_t>;
+  using ScalarLoadPolicy = CastLoadPolicy<scalar_t, acc_t>;
+  using StorePolicy = CastStoreAccumulate<scalar_t, acc_t>;
+
+  constexpr int64_t vec_stride = vacc_t::size() * sizeof(scalar_t);
+  const int64_t vec_size = size0 / vacc_t::size();
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/SumKernel.cpp:351]:
```suggestion
```

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29447187).

**Comment by ngimel** [comment]
Hm, ghimport doesn't see my changes when I import it.

**Comment by peterbell10** [comment]
@ngimel I ran `ghstack` with your change. Maybe it'll import now?

**Comment by ngimel** [comment]
I've applied changes directly to internal version, waiting for the tests to run
